Update shadow DOM and dialog element focusing to latest spec

nt1m opened the pull request:
#### 696e8c1c47e22379b8350e9228c76b595b6c54ec
<pre>
Update shadow DOM and dialog element focusing to latest spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=233320">https://bugs.webkit.org/show_bug.cgi?id=233320</a>
&lt;rdar://85753180&gt;

Reviewed by Ryosuke Niwa.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-focus-shadow-double-nested-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-focus-shadow-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::autoFocusDelegate):
(WebCore::findFocusDelegateInternal):
(WebCore::Element::findFocusDelegate):
(WebCore::Element::focus):
(WebCore::focusDelegateFromShadowHost): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::runFocusingSteps):

Canonical link: <a href="https://commits.webkit.org/252959@main">https://commits.webkit.org/252959@main</a>
</pre>

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3237,6 +3237,11 @@ static RefPtr<Element> focusDelegateFromShadowHost(ShadowRoot& target)
     return nullptr;
 }
 
+RefPtr<Element> Element::focusDelegate()
```
I think we should call this function findFocusDelegate per WebKit's coding guideline:
https://webkit.org/code-style-guidelines/#names-verb

Also, could you add the following URL in a comment for a reference?
https://html.spec.whatwg.org/multipage/interaction.html#focus-delegate

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3222,13 +3222,13 @@ static RefPtr<Element> autoFocusDelegate(ShadowRoot& target)
 
 // https://html.spec.whatwg.org/multipage/interaction.html#focus-delegate
-static RefPtr<Element> focusDelegateFromShadowHost(ShadowRoot& target)
+static RefPtr<Element> focusDelegateFromContainerNode(ContainerNode& target)
```
Why not calling this function findFocusDelegateInternal or findFocusDelegateRecursively.

rniwa commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-focus-shadow-expected.txt:
```diff
@@ -23,30 +20,24 @@ PASS show: Autofocus after, no delegatesFocus
-PASS show: Autofocus inside shadow tree, yes delegatesFocus, sibling after
-PASS showModal: Autofocus inside shadow tree, yes delegatesFocus, sibling after
+FAIL show: Autofocus inside shadow tree, yes delegatesFocus, sibling after assert_equals: expected Element node <button tabindex="-1" class="focus-me">Focusable</button> but got Element node <div></div>
+FAIL showModal: Autofocus inside shadow tree, yes delegatesFocus, sibling after assert_equals: expected Element node <button tabindex="-1" class="focus-me">Focusable</button> but got Element node <div></div>
```
What's up with these two failures?

nt1m replied:
I'll followup with Domenic to figure out if this test needs fixing.

rniwa reviewed (commented):
r=me. It's unclear whether we have a bug or the test is wrong with the failing test cases.

nt1m committed:
Update shadow DOM and dialog element focusing to latest spec
https://bugs.webkit.org/show_bug.cgi?id=233320
<rdar://85753180>

Reviewed by Ryosuke Niwa.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-focus-shadow-double-nested-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-focus-shadow-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::autoFocusDelegate):
(WebCore::findFocusDelegateInternal):
(WebCore::Element::findFocusDelegate):
(WebCore::Element::focus):
(WebCore::focusDelegateFromShadowHost): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::runFocusingSteps):

Canonical link: https://commits.webkit.org/252959@main
